content/docs/testing-recipes.md

```diff
@@ -0,0 +1,403 @@
+  act(() => {
+    for (let i = 0; i < 5; i++) {
+      button.dispatchEvent(new MouseEvent("click", { bubbles: true }));
+    }
+  });
+  expect(clicks).toBe(6);
+  expect(button.innerText).toBe("Turn on!");
+});
+```
+
+### Timers {#timers}
+
+UIs could use timer based functions like `setTimeout` to run time sensitive code. In this example, a multiple choice panel waits for a selection and advances, timing out if a selection isn't made in 5 seconds.
+
+```jsx
+function Card(props) {
+  useEffect(
+    props => {
+      const timeoutID = setTimeout(() => {
+        props.onSelect(null);
+      }, 5000);
+      return () => {
+        clearTimeout(timeoutID);
+      };
+    },
+    [props.onSelect]
+  );
+  return [1, 2, 3, 4].map(choice => {
+    <button
+      key={choice}
+      data-test-id={choice}
+      onClick={() => props.onSelect(choice)}
+    >
+      {choice}
+    </button>;
+  });
+}
+```
+
+We can write tests for this component by leveraging jest's timer mocks, and testing the different states it can be in.
+
+```jsx
+// card.test.js
+jest.useFakeTimers();
+
+let container = null
+beforeEach(() => {
+  // setup a DOM element as a render target
+  container = document.createElement('div')
+  document.body.appendChild(container)
+})
+
+afterEach(() => {
+  // cleanup on exiting
+  container.remove();
+  container = null  
+})
+
+function render(ui){
+  act(() => {
+    ReactDOM.render(ui, container);
+  });
+}
+
+it("should select null after timing out", () => {
+  const yields = [];
+  render(<Card onSelect={selection => yields.push(selection)} />);
+  act(() => {
+    jest.advanceTimersByTime(100);
+  });
+  expect(yields).toBe([]);
+  act(() => {
+    jest.advanceTimersByTime(1000);
+  });
+  expect(yields).toBe([null]);
+});
+
+it("should cleanup on being removed", () => {
+  const yields = [];
+  render(
+    <Card onSelect={selection => yields.push(selection)} />
+  );
+  act(() => {
+    jest.advanceTimersByTime(100);
+  });
+  expect(yields).toBe([]);
+  act(() => {
+    render(null);
+  });
+  act(() => {
+    jest.advanceTimersByTime(1000);
+  });
+  expect(yields).toBe([]);
+});
+
+it("should accept selections", () => {
+  const yields = [];
+  render(
+    <Card onSelect={selection => yields.push(selection)} />
+  );
+
+  act(() => {
+    container
+      .querySelector("[data-test-id=2]")
+      .dispatchEvent(new MouseEvent("click", { bubbles: true }));
+  });
+
+  expect(yields).toBe([2]);
+});
+```
+
+### Snapshot testing {#snapshot-testing}
+
+Frameworks like Jest also let you save 'snapshots' of data. Combined with react-test-renderer, we can save component structures, and verify they don't break in the future. It's typically better to make more specific assertions than to use snapshots (or at least ensure that the snapshot is small) because these kinds of tests include implementation details meaning they break easily and teams can get desensitized to snapshot breakages.
```

Comment:
>It's typically better to make more specific assertions than to use snapshots

This is a very strong recommendation. In practice snapshots might be better overall than over-specific tests that make *too* "specific" assertions. Can we relax this wording a little bit? Try to clarify pitfalls and guiding principles rather than make value judgements ("better").

Reply:
I can't tell you how many teams I've talked with who regret diving into snapshot testing. It's been a lot of teams.

I'm not opposed to changing the wording up a bit to be less strong I guess, but I think we should try to encourage effective use of snapshots and warn of the dangers of over-using snapshots.

I agree with @alexkrolick, using an inline snapshot in this example could go a long way toward helping people avoid the pitfalls naturally.

Reply:
I'll make this a *little* less sever, but I've also been thinking of recommending `toMatchInlineSnapshot`
- it's inline, so plainly visible when reading the test
- if it grows big, it's a sign that the test is losing it's efficacy, and will prompt a more focused test.